contracts/FeeCustody.sol

```diff
@@ -0,0 +1,377 @@
+     * @param _asset asset to swap from
+     * @param _amountIn amount to swap of asset
+     */
+    function _swap(address _asset, uint256 _amountIn) internal {
+      TransferHelper.safeApprove(asset, address(UNIV3_SWAP_ROUTER), _amountIn);
+
+      address[] memory _intermediaryPath = intermediaryPath[_asset];
+      address[] memory _poolFees = poolFees[_asset];
+
+      // Multiple pool swaps are encoded through bytes called a `path`.
+      // A path is a sequence of token addresses and poolFees that define the pools used in the swaps.
+      // The format for pool encoding is (tokenIn, fee, tokenOut/tokenIn, fee, tokenOut)
+      // where tokenIn/tokenOut parameter is the shared token across the pools.
+
+      bytes pathEncoding;
+      if(_intermediaryPath.length > 0){
+        pathEncoding = abi.encodePacked(_asset, _poolFees[0], _intermediaryPath[0], _poolFees[1], address(distributionToken));
+      }else{
+        pathEncoding = abi.encodePacked(_asset, _poolFees[0], address(distributionToken));
+      }
+
+      ISwapRouter.ExactInputParams memory params =
+          ISwapRouter.ExactInputParams({
+              path: pathEncoding,
+              recipient: msg.sender,
+              deadline: block.timestamp,
+              amountIn: amountIn,
+              amountOutMinimum: 0
+          });
+
+      // Executes the swap.
+      UNIV3_SWAP_ROUTER.exactInput(params);
+    }
+
+    /**
+     * @notice
+     * add asset
+     * @dev Can be called by admin
+     * @param _asset new asset
+     * @param _oracle ASSET/USD ORACLE.
+     * @param _intermediaryPath path for univ3 swap.
+     * @param _poolFee fees for asset / distributionToken.
+     * @param _isUpdate if we are updating existing asset / setting asset
+
+     * If intermediary path then pool fee between both pairs
+     * (ex: AAVE / ETH , ETH / USDC)
+     * NOTE: if intermediaryPath empty then single hop swap
+     * OR remain unchanged if asset had already been added.
+     * NOTE: MUST BE ASSET / USD ORACLE
+     * NOTE: 3000 = 0.3% fee for pool fees
+     */
+    function setAsset(address _asset, address _oracle, address[] calldata _intermediaryPath, address[] calldata _poolFees, bool _isUpdate) external onlyAdmin {
+        require(_asset != address(0), "!address(0)");
+        uint8 _pathLen = _intermediaryPath.length;
+        uint8 _swapFeeLen = _poolFees.length;
+        uint8 _assetExists = assetExists[_asset];
+
+        // We must be setting new valid oracle, or want to keep as is if one exists
+        require((_assetExists != address(0) && _oracle == address(0)) || IChainlink(oracles[_asset]).decimals() == 8, "!ASSET/USD");
+        require(_pathLen < 2, "invalid intermediary path");
+        require(((!_assetExists && _swapFeeLen > 0) || _assetExists) &&  _swapFeeLen < 2, "invalid pool fees array length");
+
+        // If not set asset
+        if(!_assetExists){
+          assets[lastAssetIdx] = _asset;
```

Comment:
There are some edge cases here that we may have to address, since `assets` is a fixed-size array. When `lastAssetIdx = 100`, line 265 will revert

Reply:
yeah it will. i am betting that we wont have more than 100 assets so that we can get static size

Reply:
changed to 1000